src/rust/engine/rule_graph/src/builder.rs

```diff
@@ -272,6 +275,36 @@ impl<R: Rule> Builder<R> {
     self.finalize(pruned_edges_graph)
   }
 
+  ///
+  /// Validate that all rules have unique RuleIds.
+  ///
+  fn validate_rule_ids(&self) -> Result<(), String> {
+    let mut invalid_rule_ids: Vec<&RuleId> = self
+      .rules
+      .values()
+      .flatten()
+      .group_by(|rule| rule.id())
+      .into_iter()
+      .filter_map(|(rule_id, group)| {
+        if group.count() > 1 {
+          Some(rule_id)
+        } else {
+          None
+        }
+      })
+      .collect();
+    match invalid_rule_ids.len() {
+      0 => Ok(()),
+      _ => {
+        invalid_rule_ids.sort();
+        Err(format!(
+          "Multiple rules have these names: {}",
```

Comment:
The message is a little awkwardly phrased, to keep it simple. But this should never happen, and certainly never happen to end-users.